pkg/controllers/leaderworkerset_controller.go

```diff
@@ -537,51 +543,101 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
 	})
 	var stsList appsv1.StatefulSetList
 	if err := r.List(ctx, &stsList, stsSelector, client.InNamespace(lws.Namespace)); err != nil {
-		return 0, 0, err
+		return nil, err
 	}
 	sortedSts := utils.SortByIndex(func(sts appsv1.StatefulSet) (int, error) {
 		return strconv.Atoi(sts.Labels[leaderworkerset.GroupIndexLabelKey])
 	}, stsList.Items, int(stsReplicas))
 
 	// Once size==1, no worker statefulSets will be created.
 	noWorkerSts := *lws.Spec.LeaderWorkerTemplate.Size == 1
-	processReplica := func(index int32) (ready bool) {
-		nominatedName := fmt.Sprintf("%s-%d", lws.Name, index)
+
+	for idx := int32(0); idx < stsReplicas; idx++ {
+		nominatedName := fmt.Sprintf("%s-%d", lws.Name, idx)
 		// It can happen that the leader pod or the worker statefulset hasn't created yet
 		// or under rebuilding, which also indicates not ready.
-		if nominatedName != sortedPods[index].Name || (!noWorkerSts && nominatedName != sortedSts[index].Name) {
-			return false
+		if nominatedName != sortedPods[idx].Name || (!noWorkerSts && nominatedName != sortedSts[idx].Name) {
+			states[idx] = replicaState{
+				ready:   false,
+				updated: false,
+			}
 		}
 
-		podTemplateHash := revisionutils.GetRevisionKey(&sortedPods[index])
-		if !(podTemplateHash == revisionKey && podutils.PodRunningAndReady(sortedPods[index])) {
-			return false
-		}
+		leaderUpdated := revisionutils.GetRevisionKey(&sortedPods[idx]) == revisionKey
+		leaderReady := podutils.PodRunningAndReady(sortedPods[idx])
 
 		if noWorkerSts {
-			return true
+			states[idx] = replicaState{
+				ready:   leaderReady,
+				updated: leaderUpdated,
+			}
+			continue
+		}
+
+		stsExists := (int(idx) < len(sortedSts) && nominatedName == sortedSts[idx].Name)
+		if !stsExists {
+			continue
 		}
 
-		stsTemplateHash := revisionutils.GetRevisionKey(&sortedSts[index])
-		return stsTemplateHash == revisionKey && statefulsetutils.StatefulsetReady(sortedSts[index])
+		workersUpdated := revisionutils.GetRevisionKey(&sortedSts[idx]) == revisionKey
+		workersReady := statefulsetutils.StatefulsetReady(sortedSts[idx])
+
+		states[idx] = replicaState{
+			ready:   leaderReady && workersReady,
+			updated: leaderUpdated && workersUpdated,
+		}
 	}
 
-	var skip bool
-	var continuousReadyReplicas, lwsUnreadyReplicas int32
+	return states, nil
+}
 
-	for index := stsReplicas - 1; index >= 0; index-- {
-		replicaReady := processReplica(index)
-		skip = skip || !replicaReady
+func calculateLWSUnreadyReplicas(states []replicaState, lwsReplicas int32) int32 {
+	var unreadyCount int32
+	for idx := int32(0); idx < min(int32(len(states)), lwsReplicas); idx++ {
```

Comment:
Is this right? Shouldn't we use lwsReplicas as the evaluation baseline？For example, lwsReplicas=10, the len(stats)=5, because the other 5 is rebuilding, then we at least have 5 unready replicas.